### PR TITLE
Support Envoy gzip compression filter

### PR DESCRIPTION
**Pull request by ilackarms**
Adds support for enabling and configuring the Envoy gzip compression filter (https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/http/gzip/v2/gzip.proto)

Users specify the `gzip` config (imported directly from Envoy) option on the Gateway `spec.httpGateway.options`.
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/2315

### CODE REVIEW

File: projects/gloo/pkg/plugins/gzip/plugin.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+func (p *Plugin) HttpFilters(_ plugins.Params, listener *v1.HttpListener) ([]plugins.StagedHttpFilter, error) {
+
+	gzipConfig := listener.GetOptions().GetGzip()
+
+	if gzipConfig == nil {
+		return nil, nil
+	}
+
+	gzipFilter, err := plugins.NewStagedFilterWithConfig(util.HealthCheck, gzipConfig, pluginStage)
```
Comment (andreasph) [inline projects/gloo/pkg/plugins/gzip/plugin.go]: it seems this should be something like `util.GZip`

File: projects/gloo/api/external/envoy/config/filter/http/gzip/v2/gzip.proto [hunk 2]
```diff
@@ -0,0 +1,83 @@
+// copied from https://raw.githubusercontent.com/envoyproxy/envoy/bd637fc7aab5de06707e3e478f507c2e7aacad75/api/envoy/config/filter/http/gzip/v2/gzip.proto
```
Comment (kdorosh) [inline projects/gloo/api/external/envoy/config/filter/http/gzip/v2/gzip.proto]: opportunity to use https://github.com/solo-io/anyvendor here? might pose problems with the go package and gogo changes, but still worth thinking about

File: projects/gloo/api/v1/options.proto [hunk 3]
```diff
@@ -82,6 +83,13 @@ message HttpListenerOptions {
     // Currently these extensions will only work if Gloo deployed using the helm flag, wasm.enabled=true
     // These require a special nightly version of envoy which is not deployed by default
     wasm.options.gloo.solo.io.PluginSource wasm = 7;
+
+    // Gzip is an HTTP option which enables Gloo to compress
+    // data returned from an upstream service upon client request.
+    // Compression is useful in situations where large payloads need to be transmitted without compromising the response time.
+    // Example:
+    //
```
Comment (kdorosh) [inline projects/gloo/api/v1/options.proto]: are we adding an example?

File: projects/gloo/pkg/plugins/gzip/plugin_test.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+						Value: 10,
+					},
+				},
+			},
+		})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(filters).To(Equal([]plugins.StagedHttpFilter{
+			plugins.StagedHttpFilter{
+				HttpFilter: &envoy_config_filter_network_http_connection_manager_v2.HttpFilter{
+					Name: "envoy.health_check",
```
Comment (andreasph) [inline projects/gloo/pkg/plugins/gzip/plugin_test.go]: `envoy.health_check` should be `envoy.gzip`

File: projects/gloo/api/external/envoy/config/filter/http/gzip/v2/gzip.proto (hunk 2 above)
Comment (scootweiss) [reply projects/gloo/api/external/envoy/config/filter/http/gzip/v2/gzip.proto]: i looked into this; currently anyvendor doesn't support protos that cannot be vendored with `go mod`. at some point we should look into supporting it for vendoring all envoy protos

File: projects/gloo/api/v1/options.proto [hunk 5]
```diff
@@ -88,7 +88,11 @@ message HttpListenerOptions {
     // data returned from an upstream service upon client request.
     // Compression is useful in situations where large payloads need to be transmitted without compromising the response time.
     // Example:
-    //
+    // ```
+    // gzip:
+    //  contentType: "application/json"
```
Comment (andreasph) [inline projects/gloo/api/v1/options.proto]:
```
contentType: 
  - "application/json"
```
isn't it?

**Comment by scootweiss** [comment]
/kick

ci flake:
```
Step #11 - "test": ------------------------------
Step #11 - "test": â€¢ Failure in Spec Setup (BeforeEach) [30.015 seconds]
Step #11 - "test": Grpc Web in memory Grpc Web Grpc [BeforeEach] works with grpc web 
Step #11 - "test": /workspace/gloo/test/e2e/grpcweb_test.go:98
Step #11 - "test": 
Step #11 - "test":   Timed out after 30.000s.
Step #11 - "test":   Expected
Step #11 - "test":       <*status.statusError | 0xc00357a820>: {
Step #11 - "test":           Code: 12,
Step #11 - "test":           Message: "",
Step #11 - "test":           Details: nil,
Step #11 - "test":           XXX_NoUnkeyedLiteral: {},
Step #11 - "test":           XXX_unrecognized: nil,
Step #11 - "test":           XXX_sizecache: 0,
Step #11 - "test":       }
Step #11 - "test":   to be nil
Step #11 - "test": 
Step #11 - "test":   /go/pkg/mod/github.com/onsi/ginkgo@v1.11.0/internal/leafnodes/runner.go:113
Step #11 - "test": 
Step #11 - "test":   Full Stack Trace
Step #11 - "test":   github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync(0xc0002ffd40, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
Step #11 - "test":   	/go/pkg/mod/github.com/onsi/ginkgo@v1.11.0/internal/leafnodes/runner.go:113 +0xe0
Step #11 - "test":   github.com/onsi/ginkgo/internal/leafnodes.(*runner).run(0xc0002ffd40, 0xc00025b608, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
Step #11 - "test":   	/go/pkg/mod/github.com/onsi/ginkgo@v1.11.0/internal/leafnodes/runner.go:64 +0x108
Step #11 - "test":   github.com/onsi/ginkgo/internal/leafnodes.(*SetupNode).Run(0xc00048a598, 0x40990c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
Step #11 - "test":   	/go/pkg/mod/github.com/onsi/ginkgo@v1.11.0/internal/leafnodes/setup_nodes.go:15 +0x88
Step #11 - "test":   github.com/onsi/ginkgo/internal/spec.(*Spec).runSample(0xc00095c690, 0x0, 0x40990c0, 0xc00010bd40)
Step #11 - "test":   	/go/pkg/mod/github.com/onsi/ginkgo@v1.11.0/internal/spec/spec.go:193 +0x2c6
Step #11 - "test":   github.com/onsi/ginkgo/internal/spec.(*Spec).Run(0xc00095c690, 0x40990c0, 0xc00010bd40)
Step #11 - "test":   	/go/pkg/mod/github.com/onsi/ginkgo@v1.11.0/internal/spec/spec.go:138 +0x168
Step #11 - "test":   github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpec(0xc000558280, 0xc00095c690, 0x0)
Step #11 - "test":   	/go/pkg/mod/github.com/onsi/ginkgo@v1.11.0/internal/specrunner/spec_runner.go:200 +0x173
Step #11 - "test":   github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpecs(0xc000558280, 0x1)
Step #11 - "test":   	/go/pkg/mod/github.com/onsi/ginkgo@v1.11.0/internal/specrunner/spec_runner.go:170 +0x217
Step #11 - "test":   github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run(0xc000558280, 0x8)
Step #11 - "test":   	/go/pkg/mod/github.com/onsi/ginkgo@v1.11.0/internal/specrunner/spec_runner.go:66 +0x141
Step #11 - "test":   github.com/onsi/ginkgo/internal/suite.(*Suite).Run(0xc00014eaf0, 0x7f2b20457050, 0xc00005f200, 0x3a060c1, 0x9, 0xc000302620, 0x1, 0x1, 0x41272e0, 0xc00010bd40, ...)
Step #11 - "test":   	/go/pkg/mod/github.com/onsi/ginkgo@v1.11.0/internal/suite/suite.go:62 +0x6a0
Step #11 - "test":   github.com/onsi/ginkgo.RunSpecsWithCustomReporters(0x409bda0, 0xc00005f200, 0x3a060c1, 0x9, 0xc0000c8ea0, 0x1, 0x1, 0xc000302100)
Step #11 - "test":   	/go/pkg/mod/github.com/onsi/ginkgo@v1.11.0/ginkgo_dsl.go:226 +0x339
Step #11 - "test":   github.com/onsi/ginkgo.RunSpecs(0x409bda0, 0xc00005f200, 0x3a060c1, 0x9, 0x1)
Step #11 - "test":   	/go/pkg/mod/github.com/onsi/ginkgo@v1.11.0/ginkgo_dsl.go:207 +0x24e
Step #11 - "test":   github.com/solo-io/gloo/test/e2e_test.TestE2e(0xc00005f200)
Step #11 - "test":   	/workspace/gloo/test/e2e/e2e_suite_test.go:41 +0x1d6
Step #11 - "test":   testing.tRunner(0xc00005f200, 0x3afe6f0)
Step #11 - "test":   	/usr/local/go/src/testing/testing.go:909 +0x19a
Step #11 - "test":   created by testing.(*T).Run
Step #11 - "test":   	/usr/local/go/src/testing/testing.go:960 +0x652
Step #11 - "test": ------------------------------
Step #11 - "test": SSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
```

File: projects/gloo/pkg/plugins/gzip/plugin.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import (
+	"github.com/pkg/errors"
+	"github.com/solo-io/solo-kit/pkg/api/v1/control-plane/util"
+
+	v1 "github.com/solo-io/gloo/projects/gloo/pkg/api/v1"
+	"github.com/solo-io/gloo/projects/gloo/pkg/plugins"
+)
+
+// filter info
+var pluginStage = plugins.AfterStage(plugins.AuthZStage)
```
[the remark this answers is not in the document]
Comment (scootweiss) [reply projects/gloo/pkg/plugins/gzip/plugin.go]: it makes sense to run gzip only after routing decision has been made. definitely after transformation. i placed in `During(Routing)` which i checked comes after transformation.

**Comment by andreasph** [comment]
tested `1.3.3-10-g34bc23ba` image build: gzip content encoding works as expected

**Pull request merged**